### PR TITLE
Improvement: Improve gcode preview

### PR DESCRIPTION
**Pull request by jane-rose**

### CODE REVIEW

File: src/app/ui/widgets/PrintingVisualizer/ModeToggleBtn.jsx [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+                            onToggleMode(value);
+                        }}
+                    >
+
+                        {
+                            GCODEPREVIEWMODES.map((item, index) => {
+                                return (
+                                    <div
+                                        key={item}
+                                        style={{ margin: '16px', marginRight: '10px' }}
```
Comment (jane-rose) [inline src/app/ui/widgets/PrintingVisualizer/ModeToggleBtn.jsx:65]: use default className

File: src/app/ui/widgets/PrintingVisualizer/VisualizerPreviewControl.jsx [hunk 2]
```diff
@@ -37,69 +40,80 @@ function useShowToggleBtn() {
...
+    const [x, setX] = useState(layerCount / (MAX - MIN));
+    useEffect(() => {
+        setX(layerCount / (MAX - MIN));
+    }, [layerCount]);
+
+    const [value, setValue] = useState([]);
+    useEffect(() => {
+        // console.log('layerRangeDisplayed v', layerRangeDisplayed[0], layerRangeDisplayed[1], value[0] * x,
+        //     (value[1] - MIN) * x, layerRangeDisplayed[0] / x,
+        //     layerRangeDisplayed[1] / x + MIN);
```
Comment (jane-rose) [inline src/app/ui/widgets/PrintingVisualizer/VisualizerPreviewControl.jsx:61]: log

File: src/app/ui/widgets/PrintingVisualizer/VisualizerPreviewControl.jsx [hunk 3]
```diff
@@ -37,69 +40,80 @@ function useShowToggleBtn() {
...
+        setValue([
+            layerRangeDisplayed[0] / x,
+            layerRangeDisplayed[1] / x + MIN
+        ]);
+    }, [layerRangeDisplayed, x]);
+
+
+    const onChangeShowLayer = throttle((v) => {
+        // console.log('v', v);
```
Comment (jane-rose) [inline src/app/ui/widgets/PrintingVisualizer/VisualizerPreviewControl.jsx:70]: log

File: src/app/workers/GcodeToBufferGeometry/gcodeBufferGeometryToObj3d.js [hunk 4]
```diff
@@ -1,22 +1,41 @@
 import * as THREE from 'three';
+// import { Line2 } from 'three/examples/jsm/lines/Line2';
```
Comment (jane-rose) [inline src/app/workers/GcodeToBufferGeometry/gcodeBufferGeometryToObj3d.js:2]: log

File: src/app/workers/GcodeToBufferGeometry/index.js [hunk 5]
```diff
@@ -111,7 +112,8 @@ const gcodeToBufferGeometry = async (func, filename, extruderColors, onProgress
                 result = {
                     bufferGeometry,
                     layerCount,
-                    bounds
+                    bounds,
+                    gcode
```
Comment (jane-rose) [inline src/app/workers/GcodeToBufferGeometry/index.js:116]: not used, remove

File: src/app/workers/gcodeToBufferGeometry.js [hunk 6]
```diff
@@ -45,7 +45,8 @@ const gcodeToBufferGeometry = async (message) => {
             typeCodes,
             toolCodes,
             layerCount,
-            bounds
+            bounds,
+            gcode
```
Comment (jane-rose) [inline src/app/workers/gcodeToBufferGeometry.js:49]: not used

File: src/app/flux/printing/index.js [hunk 7]
```diff
@@ -1518,106 +1543,267 @@ export const actions = {
         });
     },
 
+    setShowOriginalModel: (show) => (dispatch, getState) => {
+        console.log('show', show);
```
Comment (xiekun1992) [inline src/app/flux/printing/index.js:1547]: remove log

File: src/app/flux/printing/index.js [hunk 8]
```diff
@@ -651,43 +676,43 @@ export const actions = {
                 bufferGeometry.setAttribute('a_type_code', typeCodeAttribute);
                 bufferGeometry.setAttribute('a_tool_code', toolCodeAttribute);
 
-                const object3D = gcodeBufferGeometryToObj3d('3DP', bufferGeometry);
 
                 dispatch(actions.destroyGcodeLine());
+
```
Comment (xiekun1992) [inline src/app/flux/printing/index.js:681]: remove unused comments

File: src/app/flux/printing/index.js [hunk 9]
```diff
@@ -1518,106 +1543,267 @@ export const actions = {
...
+                dispatch(actions.render());
+            }
+        }
+    },
+
     setGcodeColorByRenderLineType: () => (dispatch, getState) => {
         const { gcodeLine, renderLineType } = getState().printing;
-        const uniforms = gcodeLine.material.uniforms;
-        uniforms.u_color_type.value = renderLineType ? 1 : 0;
+        // if (renderLineType) {
```
Comment (xiekun1992) [inline src/app/flux/printing/index.js:1670]: remove comments

File: src/app/flux/printing/index.js [hunk 10]
```diff
@@ -1518,106 +1543,267 @@ export const actions = {
...
+            if (!isUp && (range[0] - prevRange[0]) > 0 && (prevRange[0] - range[0]) < 1
+            && (range[1] - prevRange[1]) > 0 && (prevRange[1] - range[1]) < 1) {
+                range[0] = prevRange[0];
+                range[1] = prevRange[1];
+            }
+            dispatch(actions.updateState({
+                layerRangeDisplayed: range
+            }));
+            dispatch(actions.render());
+        }, 0)();
```
Comment (xiekun1992) [inline src/app/flux/printing/index.js:1799]: why `throttle` need to use `0`

**Review by xiekun1992 — CHANGES_REQUESTED** [review]

**Pull request merged**